File: src/TcpFlatBufferCodec.cpp
#include "TcpFlatBufferCodec.h"

#include <flatbuffers/verifier.h>

#include <stdexcept>

namespace dnf
{
std::vector<std::uint8_t> FinishTcpPayload(
    flatbuffers::FlatBufferBuilder& builder,
    Dnf::Protocol::Tcp::TcpPayload payloadType,
    flatbuffers::Offset<void> payload)
{
    if (payloadType == Dnf::Protocol::Tcp::TcpPayload_NONE ||
        payload.IsNull())
    {
        throw std::invalid_argument("TCP payload must not be empty");
    }

    const auto message = Dnf::Protocol::Tcp::CreateTcpMessage(
        builder,
        TCP_PAYLOAD_PROTOCOL_VERSION,
        payloadType,
        payload);
    Dnf::Protocol::Tcp::FinishTcpMessageBuffer(builder, message);

    return std::vector<std::uint8_t>(
        builder.GetBufferPointer(),
        builder.GetBufferPointer() + builder.GetSize());
}

const Dnf::Protocol::Tcp::TcpMessage* DecodeTcpPayload(
    const std::vector<std::uint8_t>& bytes,
    Dnf::Protocol::Tcp::TcpPayload expectedType)
{
    if (bytes.empty())
    {
        throw std::runtime_error("TCP FlatBuffer payload is empty");
    }

    flatbuffers::Verifier verifier(bytes.data(), bytes.size());
    if (!Dnf::Protocol::Tcp::VerifyTcpMessageBuffer(verifier))
    {
        throw std::runtime_error("Invalid TCP FlatBuffer payload");
    }

    const auto* message =
        Dnf::Protocol::Tcp::GetTcpMessage(bytes.data());
    if (message->protocol_version() != TCP_PAYLOAD_PROTOCOL_VERSION)
    {
        throw std::runtime_error("Unsupported TCP payload protocol version");
    }

    if (message->payload_type() != expectedType ||
        message->payload() == nullptr)
    {
        throw std::runtime_error("Unexpected TCP FlatBuffer payload type");
    }

    return message;
}
} // namespace dnf

File: client/scripts/Protocol/TcpFlatBufferCodec.cs
using System;
using System.IO;
using Google.FlatBuffers;
using TcpSchema = Dnf.Protocol.Tcp;

namespace DnfMockClient.Protocol;

public static class TcpFlatBufferCodec
{
    public const ushort ProtocolVersion = 2;

    public static byte[] FinishPayload(
        FlatBufferBuilder builder,
        TcpSchema.TcpPayload payloadType,
        int payloadOffset)
    {
        if (payloadType == TcpSchema.TcpPayload.NONE || payloadOffset == 0)
        {
            throw new ArgumentException("TCP payload must not be empty.");
        }

        Offset<TcpSchema.TcpMessage> message =
            TcpSchema.TcpMessage.CreateTcpMessage(
                builder,
                ProtocolVersion,
                payloadType,
                payloadOffset);
        TcpSchema.TcpMessage.FinishTcpMessageBuffer(builder, message);
        return builder.SizedByteArray();
    }

    public static TPayload DecodePayload<TPayload>(
        byte[] bytes,
        TcpSchema.TcpPayload expectedType)
        where TPayload : struct, IFlatbufferObject
    {
        if (bytes.Length == 0)
        {
            throw new InvalidDataException("TCP FlatBuffer payload is empty.");
        }

        var buffer = new ByteBuffer(bytes);
        if (!TcpSchema.TcpMessage.TcpMessageBufferHasIdentifier(buffer) ||
            !TcpSchema.TcpMessage.VerifyTcpMessage(buffer))
        {
            throw new InvalidDataException("Invalid TCP FlatBuffer payload.");
        }

        TcpSchema.TcpMessage message =
            TcpSchema.TcpMessage.GetRootAsTcpMessage(buffer);
        if (message.ProtocolVersion != ProtocolVersion)
        {
            throw new InvalidDataException(
                "Unsupported TCP payload protocol version.");
        }

        if (message.PayloadType != expectedType)
        {
            throw new InvalidDataException(
                "Unexpected TCP FlatBuffer payload type.");
        }

        TPayload? payload = message.Payload<TPayload>();
        if (!payload.HasValue)
        {
            throw new InvalidDataException("TCP FlatBuffer payload is missing.");
        }

        return payload.Value;
    }
}
